### DISCUSSION TITLE
How can I use CTRL + wheel to zoom in and out rather than just the wheel, and how can I make the wheel simply drag the screen instead of zooming in and out?

### DISCUSSION THREAD

**Discussion by ebrahim-mansour** [Foblex/f-flow#92] [Q&A] [answered]
Hello everyone,

I just wanted to modify a few general Canvas behaviors to resemble those found on [Miro](https://miro.com):
  - I wanted to change the **"Mouse Wheel"** action to move up, down, left, and right within the canvas, instead of zooming in and out
  - I wanted to change the **"Mouse Left Click"** action to be selecting components in the Canvas rather than moving up, down, left, or right. 
  - I also wanted to transform the cursor to a pointer when the user clicks **"Space + Mouse Left Click"** so that dragging is possible.

**Comment by siarheihuzarevich** [comment]
Hello,

Thank you for your interest in Foblex Flow. To achieve the behavior you described, you can create your own custom zoom directive. By writing a custom directive, you’ll be able to:

•	Override the mouse wheel action to pan the Canvas instead of zooming.

•	Customize the left mouse click to select components instead of moving the Canvas.

• Add logic for the “Space + Left Mouse Click” combination to enable dragging and change the cursor.

The library is flexible enough to accommodate these modifications, so most of your requirements can be fulfilled with this approach.

**Reply by ebrahim-mansour** [reply]
I tried using a directive that I had already built, but I was only able to override the wheel event; I was unable to apply the scroll or dragging events.

Additionally, the overriding only works in the blue region because the library is constructing the canvas with a width and height smaller than the screen, as in the following image.

<img width="720" alt="Screenshot 2024-12-09 at 5 51 25 PM" src="https://github.com/user-attachments/assets/a77050f8-b300-4b29-b3be-89cfa821d1ad" />

Do you mean overriding the functionality in the library at a low level and rebuilding it?

Here is my code:

`
  import { Directive, HostListener, Inject } from '@angular/core'
  import { F_ZOOM, FZoomBase } from '@foblex/flow'
  
  @Directive({
    selector: '[appCtrlZoom]',
    standalone: true,
  })
  export class CtrlZoomDirective {
    constructor(@Inject(F_ZOOM) private zoomService: FZoomBase) {}
  
    @HostListener('wheel', ['$event']) onWheel(event: WheelEvent): void {
      event.preventDefault()
      event.stopPropagation()
      if (event.ctrlKey) {
        this.zoomService['onWheel'](event)
      } else {
        console.log(`🚀 scrolling....`)
        const scrollAmount = event.deltaY > 0 ? 100 : -100 // Define scroll step
        window.scrollBy(0, scrollAmount) // Adjust the scroll position vertically
      }
    }
  }
`

**Reply by siarheihuzarevich** [reply]
Yes, I mean overriding the fDraggable and fZoom directives

**Reply by ebrahim-mansour** [reply]
Since I have never done this before, I would appreciate it if you could walk me through the steps. Thanks in advance.

**Reply by siarheihuzarevich** [reply]
I think not, I don’t have enough time to explain everything to you. But you can explore the code yourself

**Reply by ebrahim-mansour** [reply]
I simply meant how to use the overrided code and integrate it, not how to explore the source.
In any case, I appreciate your response and will take care of it.

**Reply by siarheihuzarevich** [reply]
Well, I explained to you that you need to write your own fDraggable and fZoom directives that will suit your needs

**Comment by siarheihuzarevich** [comment, marked as answer]
Hi again! Just a quick update 🙂

As of version 17.1.1, you no longer need to override fZoom or fDraggable to customize canvas interactions.

A new custom event trigger system has been introduced, which allows you to configure wheel behavior, left click, keyboard combinations (like Space + Click), and more — all without writing custom directives.

Here’s a live example of how to use it:
🔗 https://flow.foblex.com/examples/custom-event-triggers

Let me know if you have any questions!
